src/transformStateWithClones.js

```diff
@@ -7,7 +7,47 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let currentState = { ...state };
+  const result = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        currentState = addProperties(currentState, action.extraData);
+        break;
+
+      case 'removeProperties':
+        currentState = removeProperties(currentState, action.keysToRemove);
+        break;
+
+      case 'clearProperties':
+        currentState = clearProperties(currentState);
+        break;
+
+      default:
+        throw new Error(`Unknown action type: ${action.type}`);
+    }
+    result.push({ ...currentState });
+  }
+
+  return result;
+}
+
+function addProperties(state, extraData) {
+  return { ...state, ...extraData };
+}
+
+function removeProperties(state, keysToRemove) {
+  const newState = { ...state };
+
+  for (const key of keysToRemove) {
+    delete newState[key];
+  }
+
+  return newState;
 }
 
+function clearProperties(state) {
```

Comment:
``clearProperties`` function does not need a ``state`` input parameter because it always returns an empty object.
```suggestion
function clearProperties() {
```